BCC upstream_submit (temporary; for discussion only)

jcanseco opened the pull request:
## About this PR
This PR is to serve as _the_ place where we discuss the patches we wish to submit to BCC upstream. Since this PR is for discussion only, this will never actually be approved and merged. 

The intention is to start small and cherry-pick only a few commits from the _joelagnel/bcc/bcc-bpfd_ branch onto **_jcanseco/bcc/upstream_submit_** (which was branched off _iovisor/bcc/master_ and is thus up-to-date with "upstream").

From the discussion, I will make the appropriate changes, whether it be code changes or commit squashing/splitting/rewording.

Once we finish discussing the commits and are satisfied, we can continue cherry-picking more commits from _joelagnel/bcc/bcc-bpfd_ and repeat the discussion once again. 

This is therefore an iterative, ongoing process, and this PR serves as the platform where we can coordinate and document our discussions during this said process.

## Testing
We can test our upstream submission for BCC by running the BCC tools against HiKey / ShellRemote via BPFd.

However, there's a catch: since BPFd depends on BCC APIs that have been changed upstream, this requires changes to BPFd itself. The BPFd source files in _jcanseco/**bcc**/upstream_submit_ therefore need to deviate from those in _joelagnel/**bpfd**/master_.

This won't do since we want to test our BCC upstream submission against the same BPFd code we intend to submit upstream, but as it is currently, only the BPFd repository has the CMake configs required to build BPFd for arm64 (which we need if we wish to test against HiKey). 

So, we introduce another branch: _jcanseco/**bpfd**/upstream_submit_, which is a branch in the BPFd repo whose source files are kept in sync with _jcanseco/**bcc**/upstream_submit_. This allows us to build BPFd for arm64 from the same sources we intend to upstream, which we can then use to test our BCC upstream submission.

tl;dr: For testing, we run the tools against the bpfd executable built from _jcanseco/**bpfd**/upstream_submit_. All comments regarding tests that have been run for our BCC upstream submission will be about tests run with this set-up.

## Notes
- GitHub [does not show the commits in the right order](https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/) when the history has been modified using rebase / force push (which we use liberally to massage the structure of our upstream submission). If you'd like to see the right order, please checkout this branch locally and invoke `git log`.

Commit:
clang: Make libclang_frontend depend on libbcc

Libclang uses functions from libbcc, so it does depend on it.
When defining new symbols in bpf_common.cc, linker errors appear
when the same symbols are used in libclang. This is because of
incorrect linker dependency. This patch fixes the issue by making
sure the dependency is correctly tracked.

Signed-off-by: Joel Fernandes <joelaf@google.com>

jcanseco committed:
bpfd, bcc/remote: Create bcc remotes to hook up BCC to BPFd

This modifies BCC to query BPFd whenever BCC wishes to perform an
operation on a remote target device (e.g. load BPF programs, read
/proc/kallsyms, attach kprobes, etc.)

If no remote target device has been configured, BCC defaults to
performing the operations on the local system just like before.

Signed-off-by: Jazel Canseco <jcanseco@google.com>

jcanseco committed:
bcc/remote: setup.py: Add remote to setup for packaging

Signed-off-by: Jazel Canseco <jcanseco@google.com>

jcanseco committed:
bpfd: Add BPFd (BPF daemon)

This patch adds BPFd, a standalone executable designed to provide BCC the
ability to work across system and architecture boundaries.

This is done by loading the BPFd executable onto a remote target device to
have it act as a proxy for whenever a BCC tool wishes to perform an operation on
the system (e.g. load BPF programs, read /proc/kallsyms, attach kprobes, etc.).

This arrangement allows developers to have kernel sources and the LLVM stack
on a separate host machine (e.g. the development machine) instead of needing
to set all these up on the target device, thereby allowing for the drastic
reduction of space required on a target for BCC tools to run. The reduction of
the space requirement, in particular, becomes a much more critical factor
for devices that have more limited disk space (e.g. embedded devices).

In addition, the above set-up also allows developers to run clang on a
different architecture than the target's architecture, thus facilitating
cross-compilation development.

However, the natural disadvantage for cross-developers is that there is a
need to have a copy of the target's kernel sources on the host for the
above set-up to work.

For more information, please check out the README in the original
BPFd repository (https://github.com/joelagnel/bpfd) and this LWN article
explaining the purpose of and how BPFd works in more detail
(https://lwn.net/Articles/744522/)

Signed-off-by: Jazel Canseco <jcanseco@google.com>

joelagnel commented on src/cc/frontends/clang/CMakeLists.txt:
```diff
@@ -10,3 +10,4 @@ if(DEFINED BCC_BACKUP_COMPILE)
 endif()
 
 add_library(clang_frontend STATIC loader.cc b_frontend_action.cc tp_frontend_action.cc kbuild_helper.cc ../../common.cc)
+target_link_libraries(clang_frontend bcc-static)
```
Drop this commit. Its already merged upstream..

jcanseco replied:
It was removed in commit [29015d7ef003f35f2a4671d67404f9e5647d6b13](https://github.com/iovisor/bcc/commit/29015d7ef003f35f2a4671d67404f9e5647d6b13)

jcanseco replied:
Okay so I just looked into this issue. Commit 29015d7  removed the line linking `clang_frontend` to `bcc_static` because it found it unnecessary. Thankfully, it doesn't seem like it had anything to do with the [bug](https://github.com/iovisor/bcc/issues/1621) the commit was trying to resolve, and was just removed cause the author happened to pass by it and figured why not do some cleaning. This means that it is most likely safe to add back in the line that has been removed.

joelagnel replied:
Does dropping this commit altogether break anything? Guessing it does

jcanseco replied:
Yes it does. The build fails without it.

joelagnel commented on src/cc/bpfd/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+# Licensed under the Apache License, Version 2.0 (the "License")
+
+add_executable(bpfd bpfd.c base64.c remote_perf_reader.c utils.c)
+target_link_libraries(bpfd bpf-static)
```
I was thinking for the upstream port, we really don't want static, it will raise a red flag.. Can you change it to be dynamic linked for upstream?

joelagnel replied:
What would be perfect is if you can make upstream dynamic, and put the static stuff in the arm build.. so this master file has nothing static. You can use a flag for that which will be set by the ARM specific compile cmake file

jcanseco replied:
Done.

jcanseco commented:
Please let me know if you approve of the current commits. I'd like to push the next few commits for review when possible. One of these commits include the fix to close perf events.

lcp committed:
tools/syscount: convert syscall and process names to bytes array

subprocess.check_output() returns a bytes array in python3 while it
returns str in python2. To improve the compatibility, convert the
related fields to bytes arrays and print them with printb().

Signed-off-by: Gary Lin <glin@suse.com>

joelagnel commented on src/cc/frontends/clang/libbpf.h:
```diff
@@ -0,0 +1,302 @@
+extern "C" {
+#endif
+
+typedef int (*bpf_create_map_cb_t)(void *cb_cookie, void *data);
```
So this wont fly at all, it was done for prototyping but there's no reason to duplicate entire libbpf.h , upstream wont accept it. Find a way to reuse the existing libbpf.h

jcanseco replied:
Done.

joelagnel commented on src/cc/frontends/clang/b_frontend_action.cc:
```diff
@@ -790,8 +790,27 @@ bool BTypeVisitor::VisitVarDecl(VarDecl *Decl) {
+        args.max_entries = table.max_entries;
+        args.map_flags = table.flags;
+
+        /* TODO: Get return value and store in table.fd */
```
This is a bit broken and wont fly upstream - did you even try to fix it? 

Can `bpf_create_map_cb` return <= 0 ? If yes, then the below is completely broken because it'll result in a call to `bpf_create_map` for remotes which I don't think makes any sense

jcanseco replied:
Okay I see the problem here. I implemented a fix that dealt with this issue while leveraging the callback mechanism you wrote.

joelagnel commented on src/python/bcc/__init__.py:
```diff
@@ -691,12 +805,21 @@ def attach_tracepoint(self, tp=b"", tp_re=b"", fn_name=b""):
         fn = self.load_func(fn_name, BPF.TRACEPOINT)
         (tp_category, tp_name) = tp.split(b':')
+
+        # TODO: Add support for remote cbs
```
I don't think remote cbs (callbacks) exists anymore so get rid of that

jcanseco replied:
Done.

joelagnel commented on src/python/bcc/libbcc.py:
```diff
@@ -17,14 +17,15 @@
 lib.bpf_module_create_c_from_string.restype = ct.c_void_p
 lib.bpf_module_create_c_from_string.argtypes = [ct.c_char_p, ct.c_uint,
-        ct.POINTER(ct.c_char_p), ct.c_int]
+        ct.POINTER(ct.c_char_p), ct.c_int, _MAP_CB_TYPE, ct.py_object]
```
Does ctypes need the ct.py_object argument for this?

jcanseco replied:
As far as I can see, yes. This is how we pass around the reference to "self" (where "self" is the BPF object in this case). This allows us to access instance variables and member functions of the BPF object.

The only instance variable we access though is "libremote", so once I apply the commit that turns that into a class variable instead of an instance variable, we can perhaps remove the `ct.py_object` argument and even the `bpf_create_map_cookie` global variable.

jcanseco replied:
This has been removed in the latest iteration of this PR.

joelagnel commented on src/python/bcc/remote/adb.py:
```diff
@@ -0,0 +1,47 @@
+            cmd = '/data/bpfd'
+
+        # HACK: Make it work on hikey 4.9 kernel for demo purpose
+        cmd += ' -k 264519'
```
How can this ever go upstream?

jcanseco replied:
Done.

joelagnel commented on src/python/bcc/remote/libremote.py:
```diff
@@ -0,0 +1,225 @@
+        return ret[0]
+
+    def bpf_detach_kprobe(self, evname):
+        # this is hard to do since a signal needs to be delivered if
```
This issue is the same as the perf reader polling issue or it could be the same thing. Either way, why isn't this fixed/cleaned up?

jcanseco replied:
Done.

joelagnel commented on tests/test_bcc_remote.py:
```diff
@@ -0,0 +1,6 @@
+from bcc.remote import shell
+
+s = shell.ShellRemote()
+print s.send_command('testxx')
```
is this test really expected to print something? And which commit has all the smoke tests your wrote??

jcanseco replied:
I intended the tests to come later as a separate commit.

But now that you mention it, it probably is better to include the tests in this commit instead. I'll apply this change shortly.

jcanseco replied:
Done.

jcanseco commented:
Updates:
- Implemented bpf_close_perf_event_fd() on BPFd side. This is a libbpf.c function that is part of the upstream BCC API changes that rendered the leaked perf reader issue moot.
- Brought in the cmd_parser changes from the BPFd repo. This revamps the input parsing logic and fixes bugs like BPFd's inability to handle commands with < 2 args.
  - Naturally, I also removed all the extra dummy arguments in libremote that were no longer needed
- Addressed all previous PR comments.
  - Note: this includes a fix for the bpf_create_map callback issue where we weren't handling the scenario where invoking bpf_create_map() on remote returns <= 0. 
- Rebased the branch on top of iovisor/master so that we are up-to-date with upstream once again.
- Modified the commit messages to be more coherent (i.e. to have a more "story-like" flow, at least more so than before)

This is all I have for now, I will now proceed to cherry-pick the other bcc-bpfd commits.

jcanseco commented:
More updates:
- Added uprobe support, kernel symbol lookup, and userspace symbol lookup
- Changed libremote into a class variable, which resulted in a couple of important clean-ups:
  - It allowed me to remove the redundant functions get_tracepoints_libremote() and get_tracepoints_remote() which originally existed just so we can pass the libremote instance to the static method get_tracepoints()
  - It allowed me to purge bpf_create_map_cb_cookie - a considerable benefit since its usage was quite confusing
  - Simplified the get_kprobe_functions() signature since it no longer needs to be passed the libremote instance

jcanseco commented:
This PR now effectively contains all commits from bcc-bpfd, except for the ones with PRs still waiting to be approved.

fser committed:
Update INSTALL.md

Add `libfl-dev` to the dependencies on debian to prevent the following error:
```
[ 16%] Building CXX object src/cc/frontends/b/CMakeFiles/b_frontend.dir/loader.cc.o
In file included from /root/bcc/src/cc/frontends/b/parser.h:21:0,
                 from /root/bcc/src/cc/frontends/b/loader.cc:17:
/root/bcc/src/cc/frontends/b/lexer.h:22:23: fatal error: FlexLexer.h: No such file or directory
 #include <FlexLexer.h>
                       ^
compilation terminated.
```

yonghong-song committed:
Merge pull request #1657 from fser/patch-1

Update INSTALL.md

jcanseco commented:
More updates:
- Cleaned up load_func() to be clearer; it was quite all over the place before
- Removed comments that littered bpfd.c regarding libbpf function signatures
- Added a log() function to libremote so that we're not cluttering BCC with checks for the "BCC_REMOTE_DEBUG" env var

joelagnel commented on src/python/bcc/__init__.py:
```diff
@@ -142,6 +151,8 @@ class BPF(object):
     XDP_TX = 3
 
+    _libremote = None
+
```
suggest kill newline

jcanseco replied:
Done.

joelagnel commented on src/cc/bpf_common.cc:
```diff
@@ -26,7 +28,9 @@ void * bpf_module_create_b(const char *filename, const char *proto_filename, uns
 
-void * bpf_module_create_c(const char *filename, unsigned flags, const char *cflags[], int ncflags) {
+void * bpf_module_create_c(const char *filename, unsigned flags, const char *cflags[], int ncflags, bpf_create_map_cb_t map_cb) {
+  bpf_create_map_cb = map_cb;
```
Could you add a comment here, why map_cb is needed - either here or in the definition. Its because its not very clear why we're doing this. What do you think

jcanseco replied:
Done.

joelagnel reviewed (changes requested):
Some comments about the commit messages. Also please mention something about external kernel sources (cross-development) for embedded developers. Check the slides for any other disadvantages worth adding.

> /proc/kallsyms, attach kprobes, etc.).
We're not really directly reading kallsyms anymore, so don't add that here

> For the below, use less "this" and more "it". I learnt early on in my career that using "this" too much doesn't sound right

* This allows us to have kernel sources and the LLVM stack on a separate
host machine (such as the development machine) instead of having to set
these up on the target device, thereby drastically reducing the space
required on a target device to run BCC tools for it. This becomes a
much more critical factor for target devices that have more limited
disk space (e.g. embedded devices). This also allows us to run clang on a
different architecture than the target's architecture, thus facilitating
cross-compilation development.*

For more information, please check out the README in the original
BPFd repository (https://github.com/joelagnel/bpfd) and this LWN article
explaining the purpose of and how BPFd works in more detail
(https://lwn.net/Articles/744522/)

Signed-off-by: Jazel Canseco <jcanseco@google.com>

jcanseco reviewed (commented):
> We're not really directly reading kallsyms anymore, so don't add that here

This is not true. We read /proc/kallsyms to perform the kernel symbol lookups.

I also checked the slides, but they mentioned no other disadvantage other than the requirement for having a copy of the kernel sources on the host machine.

jcanseco commented:
Updates:
- Pulled in the changes from the BCC and BPFd PRs approved yesterday
- Eliminated perf_reader duplication
- Fixed issue regarding BCC's improper exiting of BPFd

lcp committed:
tools/cachetop: match bpf.ksym() with bytes arrays

bpf.ksym() now returns a bytes array, and python3 would issue TypeError
when matching the string.

Signed-off-by: Gary Lin <glin@suse.com>

yonghong-song committed:
Merge pull request #1658 from lcp/syscount_str-v2

tools/syscount: convert syscall and process names to bytes array

anakryiko committed:
Return number of CPUs with data for BPFPerfBufferTable::poll()

anakryiko committed:
Return number of CPUs with data for BPFPerfBufferTable::poll()

yonghong-song committed:
Merge pull request #1659 from lcp/cachetop-bytes

tools/cachetop: match bpf.ksym() with bytes arrays

qmonnet committed:
sync bpf compat headers with latest net-next, update doc for helpers

- Update links in doc (make them point from net-next to linux, when
  relevant).
- Fix kernel version for bpf_msg_*() helpers, 4.17 instead of 4.16, in
  doc and in header.
- Add helper bpf_bind() to documentation and headers.
- Synchronise helpers with latest net-next.

jcanseco commented:
Updates:
- Increased smoke tests and added kernel version checks since we run the tests on a ShellRemote on the host machine (and so the host machine's kernel needs to be at least a certain version in some of the tests)
- Eliminated remote_perf_reader lost_cb
- Removed "-k" flag handling in BPFd since it's no longer needed
- Fix bug where detach_tracepoint() was closing a local perf_event when it should be closing a remote one. This fixes the issue where terminating syscount.py led to crashes.
- Add support for detaching tracepoints on a remote device
- Add support for remote comm_for_pid, fixing the bug wherein syscount.py was reading the comm locally instead of remotely

nirmoy committed:
Building from source on openSUSE Tumbleweed requires the clang-devel package to be installed

Signed-off-by: Nirmoy Das <ndas@suse.de>

yonghong-song committed:
Merge pull request #1660 from anakryiko/perf_buffer_poll_return_type

Return number of CPUs with data for BPFPerfBufferTable::poll()

yonghong-song committed:
Merge pull request #1661 from qmonnet/kernelfeatures

sync bpf compat headers with latest net-next, update doc for helpers

yonghong-song committed:
Merge pull request #1662 from nirmoy/suse_changes_1

Building from source on openSUSE Tumbleweed requires the clang-devel …

yonghong-song committed:
fix compilation error with latest llvm

The latest llvm (7.0) changed the interface for
IRBuilder/CreateMemCpy as in https://reviews.llvm.org/rL328317.
This caused the compilation error like below:

  [ 30%] Built target bcc-loader-static
  /home/yhs/work/bcc/src/cc/frontends/b/codegen_llvm.cc: In member function ‘virtual ebpf::StatusTuple
  ebpf::cc::CodegenLLVM::visit_string_expr_node(ebpf::cc::StringExprNode*)’:
  /home/yhs/work/bcc/src/cc/frontends/b/codegen_llvm.cc:428:52: error: no matching function for call t$
   ‘llvm::IRBuilder<>::CreateMemCpy(llvm::Value*&, llvm::Value*&, std::basic_string<char>::size_type, $
  nt)’
     B.CreateMemCpy(ptr, global, n->val_.size() + 1, 1);
                                                    ^
  /home/yhs/work/bcc/src/cc/frontends/b/codegen_llvm.cc:428:52: note: candidates are:
  In file included from /home/yhs/work/bcc/src/cc/frontends/b/codegen_llvm.cc:31:0:
  /home/yhs/work/llvm/build/install/include/llvm/IR/IRBuilder.h:422:13: note: llvm::CallInst* llvm::IR$
  uilderBase::CreateMemCpy(llvm::Value*, unsigned int, llvm::Value*, unsigned int, uint64_t, bool, llv$
  ::MDNode*, llvm::MDNode*, llvm::MDNode*, llvm::MDNode*)
     CallInst *CreateMemCpy(Value *Dst, unsigned DstAlign, Value *Src,
             ^

Now the interfaces between 7.0 and 6.0 CreateMemCpy
are completely disjoint and we are not able to find a common
interface suitable for both.

This patch fixed the issue by separating two cases based on
llvm compiler version.

Signed-off-by: Yonghong Song <yhs@fb.com>

4ast committed:
Merge pull request #1663 from iovisor/yhs_dev

fix compilation error with latest llvm

yonghong-song committed:
permit symbol resulotion for function with size 0

The issue comes up when I investigated issue #1641.
A func symbol defined in assembly code will be size of 0,
e.g., http://git.musl-libc.org/cgit/musl/tree/src/thread/x86_64/syscall_cp.s
symbol __cp_begin.
  .text
  .global __cp_begin
  .hidden __cp_begin
  ...
  __syscall_cp_asm:

  __cp_begin:
	mov (%rdi),%eax
	test %eax,%eax

and __cp_begin cannot be traced through bcc since
symbol resolution rejects any func symbol with size 0.

This patch removed size-must-not-zero restriction so that
the symbol like __cp_begin can be traced.
Command line: trace.py -p <pid> -U '<binary_path>:__cp_begin'

Signed-off-by: Yonghong Song <yhs@fb.com>

4ast committed:
Merge pull request #1664 from iovisor/yhs_dev

permit symbol resulotion for function with size 0

nirmoy committed:
fix some spelling errors

Signed-off-by: Nirmoy Das <ndas@suse.de>

yonghong-song committed:
Merge pull request #1666 from nirmoy/fix_spell_1

fix some spelling errors

yonghong-song committed:
fix compiler warning

The patch fixed the following compiler warnings:

  /home/yhs/work/bcc/src/cc/perf_reader.c: In function ‘read_data_head’:
  /home/yhs/work/bcc/src/cc/perf_reader.c:149:3: warning: dereferencing type-punned pointer will break strict-alias
  ing rules [-Wstrict-aliasing]
     uint64_t data_head = *((volatile uint64_t *)&perf_header->data_head);
     ^
  /home/yhs/work/bcc/src/cc/perf_reader.c: In function ‘read_data_head’:
  /home/yhs/work/bcc/src/cc/perf_reader.c:149:3: warning: dereferencing type-punned pointer will break strict-alias
  ing rules [-Wstrict-aliasing]
     uint64_t data_head = *((volatile uint64_t *)&perf_header->data_head);
     ^

Declaring perf_header as volatile type will force its member read from memory,
hence avoiding a forced type conversion.

Signed-off-by: Yonghong Song <yhs@fb.com>

4ast committed:
Merge pull request #1669 from iovisor/yhs_dev

fix compiler warning

palmtenor committed:
Improve clear table API in C++

Commit:
bcc/syms: Fix shared library symbol lookup

Shared library addresses needed to be mapped what the address is expected in the
symbol table. The address offset of the running shared library may be different from
the one in the SO binary file. So we have to map it correctly in order for symbol
look up to work.

Often the address and file offset are the same so it works, however in Android this
is not the case a lot of times. Fix the issue by adjusting the offset with the
file offset from the ELF.

Test: stackcount 'sys_futex' -v

Before patch:
  ffffff800814bc14 sys_futex
  775d8e0db0       syscall
  775e149fe8       [unknown]
  775e14a598       [unknown]
  775e16a9e4       android::Parcel::validateReadData(unsigned long) const
  775e05d738       android::Thread::_threadLoop(void*)
  775d942470       __pthread_start(void*)
  775d8e54f8       __start_thread
    Binder:13353_5 [13353]

After patch:
  ffffff800814bc14 sys_futex
  78ba2e1db0       syscall
  78b97d4fe8       android::IPCThreadState::getAndExecuteCommand()
  78b97d5598       android::IPCThreadState::joinThreadPool(bool)
  78b97f59e4       [unknown]
  78b8dda738       android::Thread::_threadLoop(void*)
  78ba4ffbd0       android::AndroidRuntime::javaThreadShell(void*)
  78ba343470       __pthread_start(void*)
  78ba2e64f8       __start_thread
    Binder:14001_8 [14001]

Matches Perf output:
        ffffff800814bc14 sys_futex ([kernel.kallsyms])
                   1edb0 syscall (/system/lib64/libc.so)
                   45fe8 android::IPCThreadState::getAndExecuteCommand (/system/lib64/libbinder.so)
                   46598 android::IPCThreadState::joinThreadPool (/system/lib64/libbinder.so)
                   669e4 [unknown] (/system/lib64/libbinder.so)
                   10738 android::Thread::_threadLoop (/system/lib64/libutils.so)
                   7ebd0 android::AndroidRuntime::javaThreadShell (/system/lib64/libandroid_runtime.so)
                   80470 __pthread_start (/system/lib64/libc.so)
                   234f8 __start_thread (/system/lib64/libc.so)

Fixes: https://github.com/iovisor/bcc/issues/1628

Signed-off-by: Joel Fernandes <joelaf@google.com>
Signed-off-by: Joel Fernandes <joel@linuxinternals.org>

Commit:
Added -q option to execsnoop to quote individual arguments. This helps when working with arguments that contain spaces.

Commit:
moved quoting below argument pattern match, so that the pattern match happens on the unquoted argument

Commit:
Added -q option to the USAGE string in execsnoop. Added -q option to the manpage for execsnoop

yonghong-song committed:
Merge pull request #1672 from brtmr/master

Added -q option to execsnoop to quote individual arguments.

yonghong-song committed:
Merge pull request #1670 from joelagnel/submit-fix-1628

bcc/syms: Fix shared library symbol lookup

Commit:
Add probability option to inject

Commit:
Update documentation for 4.17, add documentation for probability option

Commit:
Add bpf_get_prandom_u32() to reference guide

yonghong-song committed:
Merge pull request #1676 from hMcLauchlan/inject-4.17

Add features/docs for inject (4.17)

yonghong-song committed:
Merge pull request #1673 from palmtenor/cleartab

Improve clear table API in C++
